### PR TITLE
Added access modifiers to config types

### LINKED ISSUE #444: Consider making the config structs opaque
Presently, quinn is configured through a handful of structs with public members. This is simple to work with, but presents a forwards-compatibility hazard, because the representation of parameters cannot be changed without making a breaking change to the interface. For example, we might want to switch all time intervals to use `Duration`. Per-field validation is also awkward to perform.

Instead, each struct could have entirely private fields and a number of setters, which could consume diverse types and even return `Result` where appropriate. If changes are made, old setters could be deprecated rather than removed wholesale.

### PR DESCRIPTION
**Pull request by TimonPost**
Partially fixed #444 

Refactored: `ServerConfig`, `EndpointConfig` and `ClientConfig`.

- Changed `pub` definition into `pub (crate)`. 
- Added set fields. `fn with_xxx() -> Self { }`
- Added get fields for transport config
- Updated code that used those config types

Use can create and modifie the config as follows: 

```rust
let endpoint = EndpointConfig::default()
     .with_reset_key(vec![])
     .with_local_cid_len(10);
....
```

### CODE REVIEW

**Comment by Ralith** [comment]
Do we need public accessors for config values? Keeping it write-only would simplify the API, and let the setters be a bit more concise as no `with_` would be needed to distinguish them.

**Comment by Ralith** [comment]
I don't think there's any need to move everything over to getters within quinn-proto, also. We probably want the fields to all be `pub(crate)`.

**Comment by TimonPost** [comment]
> Do we need public getters for config values

"Instead, each struct could have entirely private fields and a number of setters, which could consume diverse types and even return Result where appropriate." - is what you said in the other issue. 

It depends on what the use case is. If the user is going to interact with it I think it can be of use. If the user is not going to interact with it we can just as well make it `pub (crate)` because there will be less code.

>  Keeping it write-only would simplify the API

What do you mean with a write-only api?

> We probably want the fields to all be pub(crate).

If we only use configuration internally we can do this yes. If the user is coming into play we might create those get/set functions and use the same API as the user does. No need to have both the option `pub (crate)` and get/set functions. 


The motivation for tose get/set functions are that the user probably wants to change a few to no config values. It can do this by calling `Config::default().with_xxx()` or just `Config::default()` in most cases. By having `with` in the name, you make the config behave like a builder. That can be built as the user want's to. 

If you don't see any reasons for those get/set functions. Why the statement: _"Instead, each struct could have entirely private fields and a number of setters, which could consume diverse types and even return Result where appropriate."_  Do you mean only have set functions with `pub (crate)` fields in the config?

**Comment by Ralith** [comment]
> is what you said in the other issue.

Yes, *setters*, as opposed to both setters and getters. I think setters are a big win here, but I don't see the case for getters.

> If the user is going to interact with it I think it can be of use. If the user is not going to interact with it we can just as well make it pub (crate) because there will be less code.

I'm not aware of any case for the user to need to extract information from an existing configuration, so indeed less code and churn seems preferable. Setters are good, but if there're no getters, we can give them more concise names while keeping the same builder-style type signature.

File: README.md [hunk 1]
```diff
@@ -94,8 +94,8 @@ client will automatically find and trust it.
 
 ## Development
 
-The quinn-proto test suite uses simulated IO for reproducibility and to avoid
-long sleeps in certain timing-sensitive tests. If the `SSLKEYLOGFILE`
+The Quinn-proto test suite uses simulated IO for reproducibility and to avoid
```
Comment (djc) [inline README.md]: The crate is spelled with a lower-case q.

File: quinn-proto/src/endpoint.rs [hunk 2]
```diff
@@ -351,7 +351,7 @@ where
         remote: SocketAddr,
         server_name: &str,
     ) -> Result<(ConnectionHandle, Connection<S>), ConnectError> {
-        config.transport.validate()?;
+        config.transport_config().validate()?;
```
Comment (djc) [inline quinn-proto/src/endpoint.rs]: We decided we didn't want getters, right?

File: quinn-proto/src/shared.rs [hunk 3]
```diff
@@ -257,33 +270,83 @@ pub struct ServerConfig<S>
...
     S: crypto::Session,
 {
-    /// Transport configuration to use for incoming connections
-    pub transport: Arc<TransportConfig>,
-
-    /// TLS configuration used for incoming connections.
-    ///
-    /// Must be set to use TLS 1.3 only.
+    /// Transport configuration that is used for incoming connections.
+    transport_config: Arc<TransportConfig>,
```
Comment (djc) [inline quinn-proto/src/shared.rs]: I don't really like this renaming; `config.transport` seems perfectly clear to me, whereas `config.transport_config` is unnecessarily repetitive.

File: quinn-proto/src/shared.rs [hunk 4]
```diff
@@ -257,33 +270,83 @@ pub struct ServerConfig<S>
...
+    /// Mutable TLS configuration used for incoming connections.
```
Comment (djc) [inline quinn-proto/src/shared.rs]: Why did this lose the sentence about being restricted to TLS 1.3?

File: quinn-proto/src/shared.rs [hunk 5]
```diff
@@ -257,33 +270,83 @@ pub struct ServerConfig<S>
...
     pub crypto: S::ServerConfig,
+    /// Private key used for authentication.
```
Comment (djc) [inline quinn-proto/src/shared.rs]: Why rewrite this comment to include less information?

File: quinn-proto/src/shared.rs (hunk 3 above)
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: Because it was a field wrapped in `Arc` I thought it could be useful to add a getter. This to reduce redundant cloning trough out the lib.

Comment (djc) [reply quinn-proto/src/shared.rs]: Okay, but in that case please also remove all the cloning that's currently there so we can assess the value of adding this getter.

File: README.md (hunk 1 above)
Comment (TimonPost) [reply README.md]: Of course, I saw Quin also, however, a crate is not capitalized but the name is. Fil fix.

File: quinn-proto/src/endpoint.rs (hunk 2 above)
Comment (TimonPost) [reply quinn-proto/src/endpoint.rs]: Because it was a field wrapped in `Arc` I thought it could be useful to add a getter. This to reduce redundant cloning trough out the lib.

File: README.md (hunk 1 above)
Comment (djc) [reply README.md]: Yeah, the project we've usually spelled with a capital. But for crates it doesn't make much sense.

File: quinn-proto/src/endpoint.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs]: This has actually added additional unnecessary cloning. It's best to keep these things explicit.

File: quinn-proto/src/shared.rs (hunk 3 above)
Comment (Ralith) [reply quinn-proto/src/shared.rs]: I think in context the shorter name is reasonable; it's within a `...Config` struct, it has a `...Config` type, and there's a similarly named field right below it. In any case, renaming things would be better done in a separate PR.

Comment (TimonPost) [reply quinn-proto/src/shared.rs]: Okay.

File: quinn-proto/src/shared.rs [hunk 6]
```diff
@@ -257,33 +273,81 @@ pub struct ServerConfig<S>
 where
     S: crypto::Session,
 {
-    /// Transport configuration to use for incoming connections
+    /// TLS configuration used for incoming connections.
```
Comment (djc) [inline quinn-proto/src/shared.rs]: Not sure why this change -- `TransportConfig` does not contain TLS configuration.

File: quinn-proto/src/shared.rs [hunk 7]
```diff
@@ -353,13 +417,30 @@ where
 /// Default values should be suitable for most internet applications.
 #[derive(Clone, Debug, Default)]
 pub struct ClientConfig<C> {
-    /// Transport configuration to use
+    /// TLS configuration used for incoming connections.
```
Comment (djc) [inline quinn-proto/src/shared.rs]: As above.

File: quinn-proto/src/shared.rs [hunk 8]
```diff
@@ -353,13 +417,30 @@ where
...
     pub transport: Arc<TransportConfig>,
-
-    /// Cryptographic configuration to use
+    /// The TLS configuration.
     pub crypto: C,
 }
 
+impl<C> ClientConfig<C> {
+    /// Set the TLS configuration used for incoming connections.
```
Comment (djc) [inline quinn-proto/src/shared.rs]: As above.

File: quinn-proto/src/shared.rs [hunk 9]
```diff
@@ -240,6 +235,7 @@ impl Default for EndpointConfig {
 }
 
 impl EndpointConfig {
+    /// Validate if the local cid length is within the legal range.
```
Comment (Ralith) [inline quinn-proto/src/shared.rs]: Logically, this validates the entire configuration; it just happens that there's currently only one field that we check.

File: quinn-proto/src/shared.rs [hunk 10]
```diff
@@ -257,33 +273,81 @@ pub struct ServerConfig<S>
 where
     S: crypto::Session,
 {
-    /// Transport configuration to use for incoming connections
+    /// TLS configuration used for incoming connections.
     pub transport: Arc<TransportConfig>,
+    /// Mutable TLS configuration used for incoming connections.
+    pub crypto: S::ServerConfig,
```
[the remark this answers is not in the document]
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: They are used by the benchmarks, interop, tests, that do not belong to the crate, and can't reference those fields. Since getters are not allowed we have to expose those fields.

File: quinn-proto/src/shared.rs [hunk 11]
```diff
@@ -257,33 +273,81 @@ pub struct ServerConfig<S>
...
+    }
+
+    /// Set microseconds after a stateless retry token was issued for which it's considered valid.
+    pub fn with_retry_token_lifetime(&mut self, retry_token_lifetime: u64) -> &mut Self {
+        self.retry_token_lifetime = retry_token_lifetime;
+        self
+    }
 
-    /// Maximum number of incoming connections to buffer.
+    /// Set the maximum supported length for the incoming connection buffer.
```
[the remark this answers is not in the document]
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: I would like to stick to the same naming convention. Which for a set method is 'singular verb' + what the method does.  This is the convention used in most places. Do you have a proposal?

File: quinn-proto/src/shared.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/shared.rs]: Can you provide specific examples of cases outside quinn-proto where either of these fields are read from?

File: quinn-proto/src/shared.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/shared.rs]: I'm not sure I understand. This comment is discussing the wording of the comment on the line it's attached to.

[the remark this answers is not in the document]
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: I found: "of incoming connections to buffer." a bit unclear, because 1) to what buffer, do I need to know there is a buffer and does it make sense for me to know. 2) The sentence just correct English, what do you mean with "connections to buffer"

File: quinn-proto/src/shared.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: quinn\src\builders.rs:132:9
quinn\src\builders.rs:142:9
self.config.crypto.set_protocols(protocols);

File: quinn-proto/src/shared.rs [hunk 12]
```diff
@@ -248,6 +244,26 @@ impl EndpointConfig {
...
         Ok(())
     }
+
+    /// Set the local cid length.
+    ///
+    /// This is the length of connection IDs for the endpoint.
+    /// This must be either 0 or between 4 and 18 inclusive. The length of the local connection IDs
+    /// constrains the amount of simultaneous connections the endpoint can maintain. The API user is
+    /// responsible for making sure that the pool is large enough to cover the intended usage.
+    pub fn with_local_cid_len(&mut self, local_cid_len: usize) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: > I think we should omit the with_ prefix here,

If we do it here, we should do it on all methods as well.

>  It's also confusing when combined with mutating setters, since IMO "with" does not connote in-place modification.

Setters are mutable anyway, we won't have immutable setters.  Therefore we just have to keep one naming convention and use for all setters.

File: quinn-proto/src/shared.rs (hunk 4 above)
Comment (TimonPost) [reply quinn-proto/src/shared.rs]: This is moved to the setter because we should denote restrictions when setting a value. When copying it at two places we will get duplicates.

File: quinn-proto/src/shared.rs [hunk 13]
```diff
@@ -353,13 +417,30 @@ where
 /// Default values should be suitable for most internet applications.
 #[derive(Clone, Debug, Default)]
 pub struct ClientConfig<C> {
-    /// Transport configuration to use
+    /// TLS configuration used for incoming connections.
     pub transport: Arc<TransportConfig>,
-
-    /// Cryptographic configuration to use
+    /// The TLS configuration.
     pub crypto: C,
```
[the remark this answers is not in the document]
Comment (TimonPost) [reply quinn-proto/src/shared.rs]:
```
error[E0616]: field `crypto` of struct `quinn_proto::shared::ClientConfig` is private
   --> quinn\src\builders.rs:193:9
    |
193 |         self.config.crypto.add_certificate_authority(cert)?;
    |         ^^^^^^^^^^^^^^^^^^

error[E0616]: field `crypto` of struct `quinn_proto::shared::ClientConfig` is private
   --> quinn\src\builders.rs:201:9
    |
201 |         self.config.crypto.enable_keylog();
    |         ^^^^^^^^^^^^^^^^^^

error[E0616]: field `crypto` of struct `quinn_proto::shared::ClientConfig` is private
   --> quinn\src\builders.rs:213:9
    |
213 |         self.config.crypto.set_protocols(protocols);
    |         ^^^^^^^^^^^^^^^^^^

error[E0616]: field `crypto` of struct `quinn_proto::shared::ClientConfig` is private
   --> quinn\src\builders.rs:226:41
    |
226 |         let crypto = Arc::make_mut(&mut x.crypto)

error[E0616]: field `transport` of struct `quinn_proto::shared::ClientConfig` is private
  --> quinn\src\tests.rs:28:5
   |
28 |     client_config.transport = Arc::new(crate::TransportConfig {
```

File: quinn-proto/src/shared.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/shared.rs]: I see. Maybe we can refactor `ServerConfigBuilder` and `ClientConfigBuilder` to render those accesses unnecessary, but it looks like that may not be trivial. In that light I'm content to leave them public for now.

File: quinn-proto/src/shared.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/shared.rs]: In the original wording, "buffer" is a verb, meaning roughly "to collect for future processing." While some discussion of what that means in this specific case may be justified, I don't think the changed wording adds any additional information.

File: quinn-proto/src/frame.rs [hunk 14]
```diff
@@ -40,7 +40,7 @@ impl coding::Codec for Type {
     }
 }
 
-pub trait FrameStruct {
+pub(crate) trait FrameStruct {
```
Comment (Ralith) [inline quinn-proto/src/frame.rs]: I'm not sure I agree with explicitly crate-qualifying all internal item definitions, and in any case, such a change would belong in a separate PR so it can be more easily reviewed.

File: quinn-proto/src/shared.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/shared.rs]: > If we do it here, we should do it on all methods as well.

Yes, by "here" I meant "in this PR." Apologies for the unclear phrasing.

> Setters are mutable anyway, we won't have immutable setters.

I expect `with_`-prefixed methods to take `self` by value and produce a new value as a result, rather than updating a pre-existing value in-place. This is one of two common patterns for builders.

**Comment by TimonPost** [comment]
Sorry, PR is kind of a mess now, squashing went wrong and some changes slipt trough that shouldn't be in here, don't want to spend any more time on this PR.
